Improve softgpu lighting accuracy and speed

unknownbrackets opened the pull request:
This improved ingame (while targeting an enemy) of Valkyria Chronicles 3 by around 2%, generally 1-2% gains.

I'm pretty confident from hardware tests about the color blending, largely from using maxed out factors and multiplying the colors together.  I tried to improve accuracy of the factors, but they're still not always correct.  The powered ones are especially inaccurate - it's clearly calculating pow in some shortcut inaccurate way.

Aside from my new tests, this also passes a lot more of gpu/commands/light.

I'm not confident MATH3D_CALL is doing anything, as compilers have seemed unwilling to consider Vec3f an HVA, but I figured I'd drop on in the hopes it forces inlining.  I did make sure x86 wasn't crashing.

-[Unknown]

unknownbrackets committed:
softgpu: Separate calculation of S/T.

We could probably reuse, but we're not right now and it complicates the
logic.

unknownbrackets committed:
softgpu: Use common SIMD matrix multiplies.

unknownbrackets committed:
softgpu: Reuse SSE/NEON matrix code.

unknownbrackets committed:
softgpu: More closely match PSP light rounding.

unknownbrackets committed:
softgpu: Correct handling of NAN attenuation.

unknownbrackets committed:
softgpu: Correct NAN spotlight exponent/direction.

unknownbrackets committed:
softgpu: Cleanup -NAN and diffuse factor.

unknownbrackets committed:
softgpu: Retain floats in diffuse/specular.

This seems to be a bit more accurate.  Color blending seems correct now,
but the factors and especially pow results are off.

Also, normalize normal to 0, 0, 1, which seems to match results better.

hrydgard commented on GPU/Math3D.h:
```diff
@@ -876,33 +882,44 @@ float vectorGetByIndex(__m128 v) {
-// v and vecOut must point to different memory.
-inline void Vec3ByMatrix43(float vecOut[3], const float v[3], const float m[12]) {
 #if defined(_M_SSE)
+inline __m128 MATH3D_CALL Vec3ByMatrix43(__m128 x, __m128 y, __m128 z, const float m[12]) {
```
should comment here that x, y, z must contain those values filling out their respective SSE registers, not just once.

hrydgard commented on GPU/Math3D.h:
```diff
@@ -945,11 +987,77 @@ inline void Vec3ByMatrix44(float vecOut[4], const float v[3], const float m[16])
+}
+
+#if defined(_M_SSE)
+inline __m128 MATH3D_CALL Norm3ByMatrix43(__m128 x, __m128 y, __m128 z, const float m[12]) {
```
same as the other comment

hrydgard commented on GPU/Software/Lighting.cpp:
```diff
@@ -86,70 +87,83 @@ void Process(VertexData& vertex, bool hasColor) {
+				int specular_attspot = (int)ceilf(256 * 2 * att * spot * specular_factor + 1);
+				if (specular_attspot > 512)
+					specular_attspot = 512;
+				Vec4<int> lsc = Vec4<int>::FromRGBA(gstate.getSpecularColor(light));
+				Vec4<int> msc = (materialupdate & 4) ? vertex.color0 : Vec4<int>::FromRGBA(gstate.getMaterialSpecular());
```
I guess these could be computed and cached outside the loop, although I doubt it'll be a huge win, especially considering how rare specular is.. So probably not heh.

unknownbrackets replied:
I tried this and it didn't seem to help performance, though it was hard to tell.  Mostly seemed to stay the same...

-[Unknown]

unknownbrackets committed:
softgpu: Clarify internal matrix multiply usage.

unknownbrackets commented:
Hm, wait, I think I broke LBP lighting somewhere...

-[Unknown]

unknownbrackets committed:
softgpu: Avoid rsqrt path for normals.

In LittleBigPlanet, it's noticeable that the lighting is very off due to
the slight loss of accuracy - possibly due to cutoff or similar.
